File: src/main/kotlin/magicbook/gtlitecore/common/worldgen/trees/WorldGenOrangeTree.java
package magicbook.gtlitecore.common.worldgen.trees;

import magicbook.gtlitecore.api.utils.GTLiteUtility;
import magicbook.gtlitecore.api.utils.GTLiteValues;
import magicbook.gtlitecore.api.utils.functions.TriConsumer;
import magicbook.gtlitecore.common.item.GTLiteMetaItems;
import magicbook.gtlitecore.common.worldgen.features.BiomeCondition;
import magicbook.gtlitecore.common.worldgen.features.TemperatureRainfallCondition;
import net.minecraft.block.BlockLog;
import net.minecraft.block.state.IBlockState;
import net.minecraft.init.Biomes;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;

import java.util.Random;

public class WorldGenOrangeTree extends AbstractTree
{

    public static int LEAVES_COLOR = 0x76C92C;

    public WorldGenOrangeTree()
    {
        super("orange", 1);
        addCondition(new BiomeCondition(Biomes.SAVANNA, 3, 0.45));
        addCondition(new BiomeCondition(Biomes.SAVANNA_PLATEAU, 6, 0.25));
        addCondition(new TemperatureRainfallCondition(3, 1.2, 1.15, 0.0, 0.2));
    }

    @Override
    public int getMinTrunkHeight(Random random)
    {
        return 2 + random.nextInt(3);
    }

    @Override
    protected void generateLeaves(World worldIn, BlockPos.MutableBlockPos blockPos,
                                  int trunkHeight, Random random,
                                  TriConsumer<World, BlockPos, IBlockState> notifier)
    {
        BlockPos.MutableBlockPos currentYBlockPos = GTLiteUtility.copy(blockPos);
        currentYBlockPos.move(EnumFacing.UP);
        for (int i = 0; i < trunkHeight + 1; i++)
        {
            Iterable<BlockPos> iterator = BlockPos.getAllInBox(
                    currentYBlockPos.offset(EnumFacing.NORTH, 2)
                            .offset(EnumFacing.WEST, 2),
                    currentYBlockPos.offset(EnumFacing.SOUTH, 2)
                            .offset(EnumFacing.EAST, 2));
            int j = i;
            iterator.forEach(leavesPos -> {
                if (Math.abs(leavesPos.getX() - currentYBlockPos.getX())
                        + Math.abs(leavesPos.getZ() - currentYBlockPos.getZ()) < 3
                        && (j < trunkHeight
                        || ((Math.abs(leavesPos.getX() - currentYBlockPos.getX())
                        + Math.abs(leavesPos.getZ() - currentYBlockPos.getZ()) == 1
                        || random.nextInt(2) == 0))))
                    notifier.accept(worldIn, leavesPos, getNaturalLeaveState());
            });
            currentYBlockPos.move(EnumFacing.UP);
        }
        notifier.accept(worldIn, currentYBlockPos, getNaturalLeaveState());
    }

    @Override
    protected void generateTrunk(World worldIn, BlockPos.MutableBlockPos blockPos,
                                 int maxHeight, Random random,
                                 TriConsumer<World, BlockPos, IBlockState> notifier)
    {
        BlockPos.MutableBlockPos upNBlockPos = GTLiteUtility.copy(blockPos);
        for (int height = 0; height < maxHeight; ++height) {
            notifier.accept(worldIn, upNBlockPos,
                    logState.withProperty(BlockLog.LOG_AXIS, BlockLog.EnumAxis.Y));
            if (height > 0) {
                EnumFacing randomFacing = EnumFacing.byHorizontalIndex(random.nextInt(4));
                notifier.accept(worldIn, upNBlockPos.offset(randomFacing),
                        logState.withProperty(BlockLog.LOG_AXIS,
                                BlockLog.EnumAxis.fromFacingAxis(randomFacing.getAxis())));
            }
            upNBlockPos.move(EnumFacing.UP);
        }
    }

    @Override
    public int getBlockColor(IBlockState blockState, IBlockAccess worldIn,
                             BlockPos blockPos, int tintIndex) {
        return LEAVES_COLOR;
    }

    @Override
    public int getItemColor(ItemStack stack, int tintIndex)
    {
        return LEAVES_COLOR;
    }

    @Override
    protected int getMooreRadiusAtHeight(int height, int trunkHeight)
    {
        if (height == 0)
            return 0;
        if (height < trunkHeight + 1)
            return 1;
        return 0;
    }

    @Override
    public ItemStack getAppleDrop(int chance)
    {
        if (GTLiteValues.RNG.nextInt(chance / 10) == 0)
            return GTLiteMetaItems.ORANGE.getStackForm(GTLiteValues.RNG.nextInt(2) + 1);
        return ItemStack.EMPTY;
    }

    @Override
    public ItemStack getApple()
    {
        return GTLiteMetaItems.ORANGE.getStackForm();
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/common/worldgen/trees/WorldGenOrangeTree.kt
package gregtechlite.gtlitecore.common.worldgen.trees

import gregtech.api.GTValues
import gregtechlite.gtlitecore.api.extension.copy
import gregtechlite.gtlitecore.common.item.GTLiteMetaItems
import gregtechlite.gtlitecore.common.worldgen.features.BiomeCondition
import gregtechlite.gtlitecore.common.worldgen.features.TemperatureRainfallCondition
import net.minecraft.block.BlockLog
import net.minecraft.block.state.IBlockState
import net.minecraft.init.Biomes
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import java.util.*
import kotlin.math.abs

class WorldGenOrangeTree : AbstractTree("orange", 1)
{
    init
    {
        addCondition(BiomeCondition(Biomes.SAVANNA, 3, 0.45))
        addCondition(BiomeCondition(Biomes.SAVANNA_PLATEAU, 6, 0.25))
        addCondition(TemperatureRainfallCondition(3, 1.2, 1.15, 0.0, 0.2))
    }

    companion object
    {
        var LEAVES_COLOR = 0x76C92C
    }

    override fun getMinTrunkHeight(random: Random): Int
    {
        return 2 + random.nextInt(3)
    }

    override fun generateLeaves(worldIn: World,
                                blockPos: BlockPos.MutableBlockPos,
                                height: Int,
                                random: Random,
                                notifier: (World?, BlockPos?, IBlockState?) -> Unit)
    {
        val currentYBlockPos = blockPos.copy()
        currentYBlockPos.move(EnumFacing.UP)
        for (i in 0..<height + 1)
        {
            val iterator = BlockPos.getAllInBox(
                currentYBlockPos.offset(EnumFacing.NORTH, 2)
                    .offset(EnumFacing.WEST, 2),
                currentYBlockPos.offset(EnumFacing.SOUTH, 2)
                    .offset(EnumFacing.EAST, 2)
            )
            val j = i
            iterator.forEach { leavesPos ->
                if (abs(leavesPos!!.x - currentYBlockPos.getX())
                        + abs(leavesPos.z - currentYBlockPos.getZ()) < 3
                    && (j < height || ((abs(leavesPos.x - currentYBlockPos.getX())
                        + abs(leavesPos.z - currentYBlockPos.getZ()) == 1
                    || random.nextInt(2) == 0))))
                {
                    notifier(worldIn, leavesPos, naturalLeaveState)
                }
            }
            currentYBlockPos.move(EnumFacing.UP)
        }
        notifier(worldIn, currentYBlockPos, naturalLeaveState)
    }

    override fun generateTrunk(worldIn: World,
                               blockPos: BlockPos.MutableBlockPos,
                               maxHeight: Int,
                               random: Random?,
                               notifier: (World?, BlockPos?, IBlockState?) -> Unit)
    {
        val upNBlockPos = blockPos.copy()
        for (height in 0..<maxHeight)
        {
            notifier(worldIn, upNBlockPos, logState?.withProperty(BlockLog.LOG_AXIS, BlockLog.EnumAxis.Y))
            if (height > 0)
            {
                val randomFacing = EnumFacing.byHorizontalIndex(random!!.nextInt(4))
                notifier(worldIn, upNBlockPos.offset(randomFacing), logState?.withProperty(BlockLog.LOG_AXIS,
                    BlockLog.EnumAxis.fromFacingAxis(randomFacing.axis)))
            }
            upNBlockPos.move(EnumFacing.UP)
        }
    }

    override fun getBlockColor(blockState: IBlockState?,
                               worldIn: IBlockAccess?,
                               blockPos: BlockPos?,
                               tintIndex: Int) = LEAVES_COLOR

    override fun getItemColor(stack: ItemStack?, tintIndex: Int) = LEAVES_COLOR

    override fun getMooreRadiusAtHeight(height: Int, trunkHeight: Int): Int
    {
        if (height == 0) return 0
        if (height < trunkHeight + 1) return 1
        return 0
    }

    override fun getAppleDrop(chance: Int): ItemStack?
    {
        if (GTValues.RNG.nextInt(chance / 10) == 0)
            return GTLiteMetaItems.ORANGE.getStackForm(GTValues.RNG.nextInt(2) + 1
        )
        return ItemStack.EMPTY
    }

    override val apple: ItemStack?
        get() = GTLiteMetaItems.ORANGE.stackForm

}
